public/app/features/dashboard/components/TransformationsEditor/TransformationOperationRow.tsx

```diff
@@ -162,13 +216,20 @@ export const TransformationOperationRow = ({
         }}
       >
         {showFilterEditor && (
-          <TransformationFilter index={index} config={configs[index].transformation} data={data} onChange={onChange} />
+          <TransformationFilter
+            prevTransformOutput={prevOutput}
+            index={index}
+            config={configs[index].transformation}
+            data={data}
```

Comment:
Shouldn't this be just:

```suggestion
            index={index}
            config={configs[index].transformation}
            data={input}
```

Reply:
I am a bit confused on this one as well but left as is to support backwards compatibility. 

We use `data.annotations` and `data.series`. I think including `data.series` is incorrect because it will include all of the original frames and ignore transformations that combine them. I handle this by combining `data.series` with `prevOutput` but I think it should only be `prevOutput` because original frames can be non existent anymore. here is example:

if we have A, B, C queries

**Current:**
Transformation 1 - joinBy A and B
Inputs: A, B, C
Outputs: joinBy-A-B, C

Transformation 2 - joinBy  joinBy-A-B and C
Inputs: A, B, C, joinBy-A-B       <------ includes A and B but they cannot be used
Outputs: joinBy-A-B-C

----------------------------------------------------------------------------------

**if we change to `prevOutput`:**
Transformation 1 - joinBy A and B
Inputs: A, B, C
Outputs: joinBy-A-B, C

Transformation 2 - joinBy  joinBy-A-B and C
Inputs: C, joinBy-A-B       <------ does not include A and B 👍 
Outputs: joinBy-A-B-C

@leeoniya maybe there is a use case we are missing here? But sounds unintuitive to include A and B as options when we don't have those frames anymore?

Reply:
fixed in [1b3574a](https://github.com/grafana/grafana/pull/98323/commits/1b3574a76084600c1ba48ebcd54abb4b350439fb)

Reply:
Can we keep the consumed frame being available in later stages to be used as input? This enables the user to transform/join that frame with another frame later if needed with high flexibility .
 For example, in next stages joining A with D frame.

Reply:
@nimaforoughi - this PR was already merged. If you can make a sample scenario for a feature request, please file it as a new ticket.